File: Tools/SpriteHD/xbrz.py
"""Drive the xBRZ WASM scaler through one long-lived node process."""

import struct
import subprocess
from pathlib import Path

import numpy as np

HERE = Path(__file__).parent


def scale(frames, factor=2):
    """frames: list of (h, w, 4) uint8 RGBA -> list of the same, `factor` times larger."""
    if not frames:
        return []
    payload = [struct.pack("<i", len(frames))]
    for frame in frames:
        payload.append(struct.pack("<ii", frame.shape[1], frame.shape[0]))
        payload.append(np.ascontiguousarray(frame, np.uint8).tobytes())
    result = subprocess.run(["node", str(HERE / "xbrz.mjs"), str(factor)],
                            input=b"".join(payload), stdout=subprocess.PIPE, check=True).stdout

    count = struct.unpack_from("<i", result, 0)[0]
    at, out = 4, []
    for _ in range(count):
        width, height = struct.unpack_from("<ii", result, at)
        at += 8
        size = width * height * 4
        frame = np.frombuffer(result, np.uint8, size, at).reshape(height, width, 4).copy()
        at += size
        # Belt and braces for the client's `a == 0 && rgb != 0 -> a = 255` rule: the
        # pre-pass already drives this to zero, so this only guards against a frame
        # reaching here unprepared.
        frame[frame[..., 3] == 0, :3] = 0
        out.append(frame)
    return out

File: Tools/SpriteHD/xbrz.mjs
// Batch xBRZ 2x over a stream of raw RGBA frames, one process for the whole run.
//
// The native xBRZ CLIs build a 64 MB colour-distance table per process (~150 ms),
// which across the corpus's 461,902 frames is two hours of pure startup instead of
// four minutes of work -- so the WASM port is driven from one long-lived process
// (docs/sprite-upscaling-research.md, "Tooling on macOS Apple Silicon").
//
// Protocol, little-endian, on stdin and stdout:
//   int32 count, then per frame: int32 width, int32 height, width*height*4 bytes RGBA.
// Output is the same shape at the scaled size.

import { Scaler } from "@kayahr/xbrz";

const FACTOR = Number(process.argv[2] ?? 2);

const input = await new Promise((resolve, reject) => {
  const chunks = [];
  process.stdin.on("data", (c) => chunks.push(c));
  process.stdin.on("end", () => resolve(Buffer.concat(chunks)));
  process.stdin.on("error", reject);
});

const count = input.readInt32LE(0);
let at = 4;
const out = [];
const scalers = new Map(); // one Scaler per distinct source size

for (let i = 0; i < count; i++) {
  const width = input.readInt32LE(at);
  const height = input.readInt32LE(at + 4);
  at += 8;
  const bytes = width * height * 4;
  const source = new Uint8ClampedArray(input.buffer, input.byteOffset + at, bytes);
  at += bytes;

  const key = `${width}x${height}`;
  let scaler = scalers.get(key);
  if (!scaler) scalers.set(key, (scaler = new Scaler(width, height, FACTOR)));

  const header = Buffer.alloc(8);
  header.writeInt32LE(scaler.targetWidth, 0);
  header.writeInt32LE(scaler.targetHeight, 4);
  out.push(header, Buffer.from(scaler.scale(source))); // scale() reuses its buffer
}

const header = Buffer.alloc(4);
header.writeInt32LE(count, 0);
process.stdout.write(Buffer.concat([header, ...out]));
